assets/js/search-categories-metabox.js

```diff
@@ -0,0 +1,186 @@
+		 * @return array List of User selected Categories.
+		 */
+		function getSelectedCategories() {
+			// Category ID's.
+			var categories = [];
+			// Get selected post Category ID's by looping through selected checkboxes.
+			$.each( $( 'input.scm-category-list-item[type=checkbox]:checked' ), function() {
+				categories.push( $( this ).val() );
+			});
+			// Return the array of selected Category ID's.
+			return categories;
+		}
+
+		// Page Load defaults.
+		$( 'p#scm-category-add' ).hide();
+
+		// Event definitions.
+
+		// Sort Categories list so that the selected items appear at the top.
+		sortCategoriesListBySelectedItems();
+
+		// Capture the user selected Categories after page loads.
+		$( '#scm-categories-ul' ).on('click', '.scm-category-list-item', function() {
+
+			var checkBoxState = $( this ).prop( 'checked' );
+			var checkBoxVal   = $( this ).val();
+
+			/*
+			 * When user deselects, remove the item from
+			 * global selectedCategories
+			 */
+			if ( false === checkBoxState ) {
+				if ( selectedCategories.length > 0 && $.inArray( checkBoxVal, selectedCategories ) !== -1 ) {
+					selectedCategories = $.grep(selectedCategories, function( a ) {
+						return a !== checkBoxVal;
+					});
+				}
+				$( this ).removeAttr( 'checked' );
+			} else {
+				/*
+				 * When user selects, add the item to
+				 * global selectedCategories
+				 */
+				if ( selectedCategories.length === 0 || $.inArray( checkBoxVal, selectedCategories ) === -1 ) {
+					selectedCategories.push( checkBoxVal );
+				}
+				$( this ).attr('checked', 'checked');
+			}
+		});
+
+		/**
+		 * @summary Displays error message in Search Categories Metabox.
+		 *
+		 * @since 0.1.0
+		 *
+		 * @global array selectedCategories List of selected Categories ID.
+		 */
+		function displayError() {
+			var errorMessage;
+
+			if ( $.type( ajax_object.something_wrong_error_message ) != 'undefined' ) {
+				errorMessage = ajax_object.something_wrong_error_message;
+			} else {
+				errorMessage = 'Something went wrong.';
+			}
+
+			$( 'ul#scm-categories-ul' ).html( '<li>' + errorMessage  + '</li>' );
+			// Reset Selected Categories.
+			selectedCategories = [];
+		}
+
+		/*
+		 * When user types on the search box,
+		 * retrieve categories based on the search term.
+		 */
+		$( '#scm-search-categories-search' ).keyup(function() {
+
+			var searchQuery = $( this ).val();
+
+			$.ajax({
+				method: "POST",
+				url: ajax_object.ajax_url,
+				data: {
+					action    : "search_categories",
+					categories: selectedCategories,
+					query     : searchQuery,
+					security  : ajax_object.ajax_nonce,
+					post_id   : ajax_object.post_id
+				}
+			}).success(function( response ) {
+
+				var jsonResult = response.data;
+
+				if ( response.success ) {
+					if( '' != jsonResult.category_list ) {
+						$( 'ul#scm-categories-ul' ).html( jsonResult.category_list );
+					} else {
+						// Error message to display when no results are found.
+						var noResultsFound;
```

Comment:
In JavaScript it is always recommended to declare all variables at the top of the function.